### PR TITLE
[5.7] Allow daemon to stop when there is no more jobs in the queue

### PR DESCRIPTION
**Pull request by deleugpn**
Amazon SQS offers alarms that trigger when a message appears in the Queue. We can use that alarm to run a Task from AWS Elastic Container Service. If the Task has an exit code, Docker will gracefully shutdown the container.

The process I'm seeking is one that I don't need to have a container running 24/7 to work the jobs from the Queue. Instead, the container will start only when there is jobs available in the Queue and it will shutdown when there's nothing more to execute. AWS Fargate makes a pretty decent service to run containers with a limited lifespan.

The alternative would be to send a sigterm to the container (somehow) and intercept that in order to shutdown the daemon from outside. Having a command that runs until there's no more job is much more simple and clean.

I felt like the implementation was fairly simple and clean. The test code, however, is not one of my best. I had to figure out a way to avoid the explicit call to `exit` so PHPUnit doesn't end execution while still breaking out of the loop, so I had to get creative =D.

I'm also open to suggestions on alternatives that I might have not thought about.

### CODE REVIEW

File: tests/Queue/QueueWorkerTest.php [hunk 1]
```diff
@@ -403,4 +441,14 @@ public function setConnectionName($name)
...
+    public function timeout()
+    {
+        return time() + 60;
+    }
+}
+
+class LoopBreakerException extends RuntimeException
+{
 }
+
```
Comment (GrahamCampbell) [inline tests/Queue/QueueWorkerTest.php]: Only one newline at end of file please, not two.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Queue/Worker.php [hunk 2]
```diff
@@ -195,16 +195,17 @@ protected function pauseWorker(WorkerOptions $options, $lastRestart)
      * @param  \Illuminate\Queue\WorkerOptions  $options
      * @param  int  $lastRestart
      */
-    protected function stopIfNecessary(WorkerOptions $options, $lastRestart)
+    protected function stopIfNecessary(WorkerOptions $options, $lastRestart, $job = null)
```
Comment (lucasmichot) [inline src/Illuminate/Queue/Worker.php]: Missing $job param docblock

File: src/Illuminate/Queue/WorkerOptions.php [hunk 3]
```diff
@@ -57,13 +64,14 @@ class WorkerOptions
      * @param  bool  $force
      * @return void
      */
-    public function __construct($delay = 0, $memory = 128, $timeout = 60, $sleep = 3, $maxTries = 0, $force = false)
+    public function __construct($delay = 0, $memory = 128, $timeout = 60, $sleep = 3, $maxTries = 0, $force = false, $stopOnEmptyQueue = false)
```
Comment (lucasmichot) [inline src/Illuminate/Queue/WorkerOptions.php]: same, missing docblock

**Comment by taylorotwell** [comment]
How is the "stop when empty" option going to be populated. You never populate it from a console command, etc. in your PR.

**Comment by deleugpn** [comment]
@taylorotwell  I assumed that I could write a new custom command to do that myself, but I can also add that to the existing `queue:work` if you want.

**Comment by deleugpn** [comment]
Demo: Demo

**Pull request merged**